index.ts

```diff
@@ -0,0 +1,251 @@
+  * @property  nextNetworkNonce - The next nonce suggested by the eth_getTransactionCount method.
+  * @property  highestSuggested - The maximum between the other two, the number returned.
+  * @property  local - Nonce details derived from pending transactions and highestSuggested
+  * @property  network - Nonce details from the eth_getTransactionCount method
+*/
+export interface NonceDetails {
+  params: {
+    highestLocallyConfirmed: number;
+    nextNetworkNonce: number;
+    highestSuggested: number;
+  };
+  local: HighestContinuousFrom;
+  network: NetWorkNextNonce;
+}
+
+/**
+  * @property  nextNonce - The highest of the nonce values derived based on confirmed and pending transactions and eth_getTransactionCount method
+  * @property  nonceDetails - details of nonce value derivation.
+  * @property  releaseLock
+*/
+export interface NonceLock{
+  nextNonce: number;
+  nonceDetails: NonceDetails;
+  releaseLock: () => void;
+}
+
+/**
+  * @property  name - The name for how the nonce was calculated based on the data used
+  * @property  nonce - The next nonce value suggested by the eth_getTransactionCount method.
+  * @property  blockNumber - The latest block from the network
+  * @property  baseCount - Transaction count from the network suggested by eth_getTransactionCount method
+*/
+export interface NetWorkNextNonce{
+  name: string;
+  nonce: number;
+  details: {
+    blockNumber: string;
+    baseCount: number;
+  };
+}
+
+/**
+  @property name- The name for how the nonce was calculated based on the data used
+  @property nonce - the next suggested nonce
+  @property details{startPoint, highest} - the provided starting nonce that was used and highest derived from it (for debugging)
+*/
+export interface HighestContinuousFrom{
+  name: string;
+  nonce: number;
+  details: {
+    startPoint: number;
+    highest: number;
+  };
+}
+export interface Transactions {
+  status: string;
+  history: [Record<string, unknown>];
+  txParams: {
+    from: string;
+    gas: string;
+    value: string;
+    nonce: string;
+  };
+}
+
+class NonceTracker {
+  private provider: Record<string, unknown>;
+
+  private blockTracker: any;
+
+  private ethQuery: any;
+
+  private getPendingTransactions: ((address: string) => Transactions[]);
+
+  private getConfirmedTransactions: ((address: string) => Transactions[]);
+
+  private lockMap: Record<string, unknown>;
+
+  constructor(opts: NonceTrackerOptions) {
+    this.provider = opts.provider;
+    this.blockTracker = opts.blockTracker;
+    this.ethQuery = new EthQuery(opts.provider);
+    this.getPendingTransactions = opts.getPendingTransactions;
+    this.getConfirmedTransactions = opts.getConfirmedTransactions;
+    this.lockMap = {};
+  }
+
+  /**
+    @returns {Promise<Record<string, unknown>>} with the key releaseLock (the gloabl mutex)
+  */
+  async getGlobalLock(): Promise<Record<string, unknown>> {
```

Comment:
We can tighten up the return type (I'm not sure if this is completely syntactically valid, but something like this):
```suggestion
  async getGlobalLock(): Promise<{ releaseLock: () => void }> {
```

Reply:
I updated the code with this change and it did not throw any error. all my test cases passed as well... So I think might be this is alright!! ?